---
title: "R Notebook"
output: html_notebook
---

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code.

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Ctrl+Shift+Enter*.

```{r}
# if (!requireNamespace("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
#     
# BiocManager::install(c("SingleR","limma","cmapR","celldex"))
# # 
# packageVersion("Matrix")
# BiocManager::install(c("BiocManager"))
# install.packages('Matrix')
# install.packages('Seurat')
# devtools::install_version("Matrix", version = "1.6.4")
```

```{r}
# install.packages('devtools')
# devtools::install_github("lanagarmire/Asgard")
```

```{r}
library('Asgard')

#Please replace Your_local_path with your real local folder

# PrepareReference(cell.info="/data/jp2079/KST_Data/AsgardData/GSE70138_Broad_LINCS_cell_info_2017-04-28.txt",
#                  gene.info="/data/jp2079/KST_Data/AsgardData/GSE70138_Broad_LINCS_gene_info_2017-03-06.txt",
#                  GSE70138.sig.info = "/data/jp2079/KST_Data/AsgardData/GSE70138_Broad_LINCS_sig_info_2017-03-06.txt",
#                  GSE92742.sig.info = "/data/jp2079/KST_Data/AsgardData/GSE92742_Broad_LINCS_sig_info.txt",
#                  GSE70138.gctx = "/data/jp2079/KST_Data/AsgardData/GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328_2017-03-06.gctx",
#                  GSE92742.gctx = "/data/jp2079/KST_Data/AsgardData/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx",
#                  Output.Dir = "/data/jp2079/KST_Data/AsgardData/DrugReference/"
# )
```

```{r}
library('Seurat')

# Load cells' cell type annotations for GSE113197
cell_types_file <- paste0(
  "https://raw.githubusercontent.com/lanagarmire/",
  "Single-cell-drug-repositioning/master/Drug/Normal_celltype.txt"
)

cell_types <- read.table(file=cell_types_file, header=TRUE, check.names=FALSE)

# Cell type of interest
cell_types_names <- c(
  	"Luminal_L2_epithelial_cells", "Luminal_L1.1_epithelial_cells", 
    "Luminal_L1.2_epithelial_cells", "Basal_epithelial_cells"
)

# Load normal sample Ind5 from GSE113197 dataset 
data <- read.table(file="/data/jp2079/KST_Data/AsgardData/GSE113197_RAW/GSM3099847_Ind5_Expression_Matrix.txt", header=TRUE, check.names=FALSE)
row.names(data) <- data[, 1]
data <- data[, -1]
celltype2 <- subset(cell_types, sample=="Ind5" & celltype %in% cell_types_names)
common <- intersect(colnames(data), rownames(celltype2))
data <- data[, common]

Epithelial2 <- CreateSeuratObject(counts=data, project="Epithelial", min.cells=3,
min.features=200, meta.data=data.frame(celltype2,cell=colnames(data),type="Normal"))

#Load normal sample Ind6 from GSE113197 dataset
data <- read.table(file="/data/jp2079/KST_Data/AsgardData/GSE113197_RAW/GSM3099848_Ind6_Expression_Matrix.txt", header=TRUE, check.names=FALSE)
row.names(data) <- data[, 1]
data <- data[, -1]
celltype3 <- subset(cell_types, sample=="Ind6" & celltype %in% cell_types_names)
common <- intersect(colnames(data), rownames(celltype3))
data<-data[,common]
Epithelial3 <- CreateSeuratObject(counts = data, project = "Epithelial", min.cells = 3, min.features = 200,meta.data=data.frame(celltype3,cell=colnames(data),type="Normal"))

#Load normal sample Ind7 from GSE113197 dataset
data<-read.table(file="/data/jp2079/KST_Data/AsgardData/GSE113197_RAW/GSM3099849_Ind7_Expression_Matrix.txt",header = T,check.names=FALSE)
row.names(data)<-data[,1]
data<-data[,-1]
celltype4 <- subset(cell_types, sample=="Ind7" & celltype %in% cell_types_names)
common <- intersect(colnames(data), rownames(celltype4))
data<-data[,common]
Epithelial4 <- CreateSeuratObject(counts = data, project = "Epithelial", min.cells = 3, min.features = 200,meta.data=data.frame(celltype4, cell=colnames(data),type="Normal"))

#Load cancer sample PDX110 from GSE123926 dataset
TNBC_PDX.data<- Read10X(data.dir = "/data/jp2079/KST_Data/AsgardData/GSE123926_RAW/GSM3516947_PDX110")
TNBC.PDX2 <- CreateSeuratObject(counts = TNBC_PDX.data, project = "TNBC", min.cells = 3, min.features = 200, meta.data=data.frame(row.names=colnames(TNBC_PDX.data), cell=colnames(TNBC_PDX.data), sample="PDX-110",type="TNBC.PDX"))

#Load cancer sample PDX322 from GSE123926 dataset
TNBC_PDX.data<- Read10X(data.dir = "/data/jp2079/KST_Data/AsgardData/GSE123926_RAW/GSM3516948_PDX322")
TNBC.PDX3 <- CreateSeuratObject(counts = TNBC_PDX.data, project = "TNBC", min.cells = 3, min.features = 200, meta.data=data.frame(row.names=colnames(TNBC_PDX.data), cell=colnames(TNBC_PDX.data), sample="PDX-332",type="TNBC.PDX"))

```

```{r}
SC.list <- list(
	TNBC.PDX2 = TNBC.PDX2,
	TNBC.PDX3 = TNBC.PDX3,
	Epithelial2 = Epithelial2,
	Epithelial3 = Epithelial3,
	Epithelial4 = Epithelial4
)
CellCycle = TRUE #Set it TRUE if you want to do Cell Cycle Regression
anchor.features=2000

for (i in 1:length(SC.list)) 
  {
  SC.list[[i]] <- NormalizeData(SC.list[[i]], verbose = FALSE)
  SC.list[[i]] <- FindVariableFeatures(SC.list[[i]], selection.method = "vst", nfeatures = anchor.features, verbose = FALSE)
  }
SC.anchors <- FindIntegrationAnchors(object.list = SC.list, anchor.features = anchor.features, dims = 1:15)
SC.integrated <- IntegrateData(anchorset = SC.anchors, dims = 1:15)
save(SC.integrated, file = "SC_integrated_1.RData")
DefaultAssay(SC.integrated) <- "integrated"
    
if (CellCycle)
  {
  ##Cell Cycle Regression
  s.genes <- cc.genes$s.genes
  g2m.genes <- cc.genes$g2m.genes
  SC.integrated <- CellCycleScoring(SC.integrated, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
  SC.integrated <- ScaleData(SC.integrated, vars.to.regress = c("S.Score", "G2M.Score"), features = rownames(SC.integrated))
  SC.integrated <- RunPCA(SC.integrated, npcs = 15, verbose = FALSE)
  } else{
  ##Run the standard workflow for visualization and clustering
  SC.integrated <- ScaleData(SC.integrated, verbose = FALSE)
  SC.integrated <- RunPCA(SC.integrated, npcs = 15, verbose = FALSE)
  }
##t-SNE and Clustering
SC.integrated <- RunUMAP(SC.integrated, reduction = "pca", dims = 1:15)
SC.integrated <- FindNeighbors(SC.integrated, reduction = "pca", dims = 1:15)
SC.integrated <- FindClusters(SC.integrated, algorithm = 1, resolution = 0.4)
save(SC.integrated, file = "SC_integrated_2.RData")

##Cell Type Annotation, set by.CellType=TRUE if you want to annotate cell  type.
by.CellType=FALSE
if(by.CellType == TRUE){
  data <- as.matrix(SC.integrated@assays$RNA@data)
  hpca.se <- HumanPrimaryCellAtlasData()
  pred.hpca <- SingleR(test = data, ref = hpca.se, assay.type.test=1, labels = hpca.se$label.main)
  cell.label <- data.frame(row.names = row.names(pred.hpca),celltype=pred.hpca$labels)
  if(length(SC.integrated@meta.data$celltype)>0){
    SC.integrated@meta.data$celltype <- cell.label$celltype
     } else{
       SC.integrated@meta.data <- cbind(SC.integrated@meta.data,cell.label)
       }
  new.cells <- data.frame()
  for(i in unique(SC.integrated$seurat_clusters)){
    sub.data <- subset(SC.integrated,seurat_clusters==i)
    temp <- table(sub.data@meta.data$celltype)
    best.cell <- names(which(temp==temp[which.max(temp)]))
    cells.temp <- data.frame(cell.id=row.names(sub.data@meta.data),celltype=best.cell)
    new.cells <- rbind(new.cells,cells.temp)
    }
  cell.meta <- SC.integrated@meta.data
  cell.id <- rownames(cell.meta)
  row.names(new.cells) <- new.cells[,1]
  new.cells <- new.cells[cell.id,]
  SC.integrated@meta.data$celltype <- new.cells$celltype
  } else{
    SC.integrated@meta.data$celltype <- paste0("C",as.numeric(SC.integrated@meta.data$seurat_clusters))
    }

#Change sample names
sample<-SC.integrated@meta.data$sample
sample[which(sample=="Ind5")]<-"Normal1"
sample[which(sample=="Ind6")]<-"Normal2"
sample[which(sample=="Ind7")]<-"Normal3"
SC.integrated@meta.data$sample<-sample

save(SC.integrated, file = "/data/jp2079/KST_Data/SC_integrated.RData")

#Visualize alignment result
DimPlot(SC.integrated, reduction = "umap", split.by = "sample",group.by = "celltype")
```

```{r}
#Case sample names
Case=c("PDX-110","PDX-332")

#Control sample names
Control=c("Normal1","Normal2","Normal3")
min.cells = 3

#Get differential gene expression profiles for every cell type (or cluster if without annotation) from Limma
library('limma')
DefaultAssay(SC.integrated) <- "RNA"
set.seed(123456)
Gene.list <- list()
C_names <- NULL
for(i in unique(SC.integrated@meta.data$celltype)){
     Idents(SC.integrated) <- "celltype"
     c_cells <- subset(SC.integrated, celltype == i)
     Idents(c_cells) <- "type"
     Samples=c_cells@meta.data
     Controlsample <- row.names(subset(Samples,sample %in% Control))
     Casesample <- row.names(subset(Samples,sample %in% Case))
     if(length(Controlsample)> min.cells & length(Casesample)> min.cells){
      expr <- as.matrix(c_cells@assays$integrated$scale.data)
      new_expr <- as.matrix(expr[,c(Casesample,Controlsample)])
      new_sample <- data.frame(Samples=c(Casesample,Controlsample),type=c(rep("Case",length(Casesample)),rep("Control",length(Controlsample))))
      row.names(new_sample) <- paste(new_sample$Samples,row.names(new_sample),sep="_")
      expr <- new_expr
      # bad <- which(rowSums(expr>0)<3)
      # expr <- expr[-bad,]
      # expr <- expr[apply(expr, 1, function(x) sum(x > 0.00001)) >= 3, ]
      mm <- model.matrix(~0 + type, data = new_sample)
      fit <- lmFit(expr, mm)
      contr <- makeContrasts(typeCase - typeControl, levels = colnames(coef(fit)))
      tmp <- contrasts.fit(fit, contrasts = contr)
      tmp <- eBayes(tmp)
      C_data <- topTable(tmp, sort.by = "P",n = nrow(tmp))
      C_data_for_drug <- data.frame(row.names=row.names(C_data),score=C_data$t,adj.P.Val=C_data$adj.P.Val,P.Value=C_data$P.Value)
      Gene.list[[i]] <- C_data_for_drug
      C_names <- c(C_names,i)
     }
}
names(Gene.list) <- C_names
save(Gene.list, file = "/data/jp2079/KST_Data/Gene_list.RData")
# 
# #Get differential genes from Seurat (Wilcoxon Rank Sum test)
# library('Seurat')
# DefaultAssay(SC.integrated) <- "RNA"
# set.seed(123456)
# Gene.list <- list()
# C_names <- NULL
# for(i in unique(SC.integrated@meta.data$celltype)){
#   Idents(SC.integrated) <- "celltype"
#   c_cells <- subset(SC.integrated, celltype == i)
#   Idents(c_cells) <- "type"
#   C_data <- FindMarkers(c_cells, ident.1 = "TNBC.PDX", ident.2 = "Normal")
#   C_data_for_drug <- data.frame(row.names=row.names(C_data),score=C_data$avg_logFC,adj.P.Val=C_data$p_val_adj,P.Value=C_data$p_val) ##for Seurat version > 4.0, please use avg_log2FC instead of avg_logFC
#   Gene.list[[i]] <- C_data_for_drug
#   C_names <- c(C_names,i)
# }
# names(Gene.list) <- C_names
# 
# #Get differential genes from DESeq2 method
# library('Seurat')
# DefaultAssay(SC.integrated) <- "RNA"
# set.seed(123456)
# Gene.list <- list()
# C_names <- NULL
# for(i in unique(SC.integrated@meta.data$celltype)){
#   Idents(SC.integrated) <- "celltype"
#   c_cells <- subset(SC.integrated, celltype == i)
#   Idents(c_cells) <- "type"
#   C_data <- FindMarkers(c_cells, ident.1 = "TNBC.PDX", ident.2 = "Normal", test.use = "DESeq2")
#   C_data_for_drug <- data.frame(row.names=row.names(C_data),score=C_data$avg_logFC,adj.P.Val=C_data$p_val_adj,P.Value=C_data$p_val) ##for Seurat version > 4.0, please use avg_log2FC instead of avg_logFC
#   Gene.list[[i]] <- C_data_for_drug
#   C_names <- c(C_names,i)
# }
# names(Gene.list) <- C_names
# 
# #Get differential genes from EdgeR
# library('edgeR')
# Case=c("PDX-110","PDX-332")
# Control=c("Normal1","Normal2","Normal3")
# DefaultAssay(SC.integrated) <- "RNA"
# set.seed(123456)
# min.cells=3 # The minimum number of cells for a cell type. A cell type is omitted if it has less cells than the minimum number.
# Gene.list <- list()
# C_names <- NULL
# for(i in unique(SC.integrated@meta.data$celltype)){
#   Idents(SC.integrated) <- "celltype"
#   c_cells <- subset(SC.integrated, celltype == i)
#   Idents(c_cells) <- "type"
#   Samples=c_cells@meta.data
#   Controlsample <- row.names(subset(Samples,sample %in% Control))
#   Casesample <- row.names(subset(Samples,sample %in% Case))
#   if(length(Controlsample)>min.cells & length(Casesample)>min.cells){
#     expr <- as.matrix(c_cells@assays$RNA@data)
#     new_expr <- as.matrix(expr[,c(Casesample,Controlsample)])
#     new_sample <- data.frame(Samples=c(Casesample,Controlsample),type=c(rep("Case",length(Casesample)),rep("Control",length(Controlsample))))
#     row.names(new_sample) <- paste(new_sample$Samples,row.names(new_sample),sep="_")
#     expr <- new_expr
#     bad <- which(rowSums(expr>0)<3)
#     expr <- expr[-bad,]
#     group <- new_sample$type
#     dge <- DGEList(counts=expr, group=group)
#     group_edgeR <- factor(group,levels = c("Control","Case"))
#     design <- model.matrix(~ group_edgeR)
#     dge <- estimateDisp(dge, design = design)
#     fit <- glmFit(dge, design)
#     res <- glmLRT(fit)
#     C_data <- res$table
#     C_data_for_drug <- data.frame(row.names=row.names(C_data),score=C_data$logFC,adj.P.Val=p.adjust(C_data$PValue,method = "BH"),P.Value=C_data$PValue)
#     Gene.list[[i]] <- C_data_for_drug
#     C_names <- c(C_names,i)
#   }
# }
# names(Gene.list) <- C_names
```

```{r}
kst_shap <- read.csv('/data/jp2079/KST_Data/shap_result_in_selected_celltypes.csv')

DC1_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/DC1_shap_deg.csv',row.names = 'Gene')
colnames(DC1_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
DC1_deg <- subset(DC1_deg, select = c('score','adj.P.Val','P.Value'))

DC2_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/DC2_shap_deg.csv',row.names = 'Gene')
colnames(DC2_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
DC2_deg <- subset(DC2_deg, select = c('score','adj.P.Val','P.Value'))

dM1_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/dM1_shap_deg.csv',row.names = 'Gene')
colnames(dM1_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
dM1_deg <- subset(dM1_deg, select = c('score','adj.P.Val','P.Value'))

dM2_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/dM2_shap_deg.csv',row.names = 'Gene')
colnames(dM2_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
dM2_deg <- subset(dM2_deg, select = c('score','adj.P.Val','P.Value'))

dM3_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/dM3_shap_deg.csv',row.names = 'Gene')
colnames(dM3_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
dM3_deg <- subset(dM3_deg, select = c('score','adj.P.Val','P.Value'))

dNKp_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/dNK_p_shap_deg.csv',row.names = 'Gene')
colnames(dNKp_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
dNKp_deg <- subset(dNKp_deg, select = c('score','adj.P.Val','P.Value'))

dNK1_deg <- read.csv('/data/jp2079/KST_Data/shap_deg/dNK1_shap_deg.csv',row.names = 'Gene')
colnames(dNK1_deg) <-c("Feature.Importance.Value","score","logfoldchanges","P.Value","adj.P.Val","pct_nz_group","pct_nz_reference","max_pct","diff_pct","cell_type","layer")
dNK1_deg <- subset(dNK1_deg, select = c('score','adj.P.Val','P.Value'))

Gene.list.kst <- list('C1'=DC1_deg, 'C2'=DC2_deg, 'C3'=dM1_deg, 'C4'=dM2_deg, 'C5'=dM3_deg, 'C6'=dNKp_deg, 'C7'=dNK1_deg)
```


```{r}
library('Asgard')
library('Seurat')
library(cmapR)

tissue='ovary'

#Load tissue specific drug reference produced by PrepareReference function as mentioned above. Please select proper tissue accroding to the disease.
my_gene_info<-read.table(file=sprintf("/data/jp2079/KST_Data/AsgardData/DrugReference/%s_gene_info.txt",tissue),sep="\t",header = T,quote = "")
my_drug_info<-read.table(file=sprintf("/data/jp2079/KST_Data/AsgardData/DrugReference/%s_drug_info.txt",tissue),sep="\t",header = T,quote = "")
drug.ref.profiles = GetDrugRef(drug.response.path = sprintf('/data/jp2079/KST_Data/AsgardData/DrugReference/%s_rankMatrix.txt',tissue),
                               probe.to.genes = my_gene_info, 
                               drug.info = my_drug_info)
```



```{r}
#Repurpose mono-drugs for every cell type                               
Drug.ident.res = GetDrug(gene.data = Gene.list, 
                        drug.ref.profiles = drug.ref.profiles, 
                        repurposing.unit = "drug", 
                        connectivity = "negative", 
                        drug.type = "all")

Drug.ident.res.kst = GetDrug(gene.data = Gene.list.kst,
                        drug.ref.profiles = drug.ref.profiles,
                        repurposing.unit = "drug",
                        connectivity = "negative",
                        drug.type = "all")
```
```{r}
##Combine P-value
CombineP = function (p){
  keep <- (p > 0) & (p <= 1)
  invalid <- sum(1L * keep) < 2
  if (invalid) {
    warning("Must have at least two valid p values")
    res <- list(chisq = NA_real_, df = NA_integer_, p = NA_real_, 
      validp = p[keep])
  }
  else {
    lnp <- log(p[keep])
    chisq <- (-2) * sum(lnp)
    df <- 2 * length(lnp)
    if (length(lnp) != length(p)) {
      warning("Some studies omitted")
    }
    res <- pchisq(chisq,df, lower.tail = FALSE)
  }
  return(res)
}

#' @title Calculate drug score
#' @description The drug score is a comprehensive estimation of drug therapeutic 
#' effects using all or a selected set of clusters. 
#' @details This function calculates drug score using cellular proportion of 
#' clusters, the significance of reversal in DEGs' expressions, and the ratio of 
#' the reversed genes. 
#' @param cell_metadata A data.frame of cell metadata. It must have a column 
#' named 'cluster' indicating which cluster cells belong, and a column named 
#' 'sample' indicating which sample cells belong. 
#' @param cluster_degs A list of differential gene expression profiles for 
#' each cluster.
#' @param cluster_drugs Drug repurposing result from GetDrug function.
#' @param tissue Reference tissue. If one used 'lung_rankMatrix.txt' in 
#' GetDrugRef function, then the Reference tissue is lung. Please use " " 
#' instead of "-" in tissue name. For example, while 
#' 'haematopoietic-and-lymphoid-tissue' is the prefix of the drug reference 
#' files, the corresponding tissue name is "haematopoietic and lymphoid tissue".
#' @param gse70138_gctx_path The gctx file contains drug responses from GSE70138 
#' dataset (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE70138).
#' @param gse92742_gctx_path The gctx file contains drug responses from GSE92742 
#' dataset (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE92742)..
#' @param clusters Select which clusters (cell types) to be used for drug score 
#' estimation. By default, it uses all clusters.
#' @param case A vector containing case sample names.
#' @param fda_drugs_only logical; if TRUE, will only return FDA-approved drugs, 
#' else, will return all drugs/compounds.
#' @return A data frame of drug score, P-value and FDR.
#' @export
#' @import cmapR
#' 

DrugScore_jaesub <- function(cell_metadata, cluster_degs, cluster_drugs, tissue,
					  gse70138_gctx_path, gse92742_gctx_path, 
					  clusters = NULL, case = NULL, fda_drugs_only = TRUE) {

  # Subset input data to the set of clusters we are interested in 
  if (length(clusters) > 0) {
  	clusters = intersect(clusters, unique(cell_metada$cluster))
    	cell_metadata = subset(cell_metadata, cluster %in% clusters)
    	cluster_drugs = cluster_drugs[clusters]
    	cluster_degs = cluster_degs[clusters]
  }
  
  # Calculate cluster proportions in diseased tissue
  if (length(case) > 0) {
    	cell_metadata <- subset(cell_metadata, sample %in% case)
  }
  clustering <- cell_metadata$cluster
  cluster_sizes <- table(clustering)
  # cluster_sizes <- cluster_sizes[which(cluster_sizes > 3)]
  cluster_prop <- round(100*cluster_sizes/nrow(cell_metadata), 2) 
  
  # Combine cluster drugs into a single data frame
  drug_list <- data.frame()
  for (i in names(cluster_drugs)) {
  	ith_cluster_drugs <- cluster_drugs[[i]]
  drug_names <- ith_cluster_drugs$Drug.name
    	ith_cluster_drugs <- ith_cluster_drugs[!duplicated(drug_names), ]
  
  # Subset to FDA drugs
    	if (fda_drugs_only) {
    		drug_names <- intersect(drug_names, FDA.drug)
    	}
  
    	if (length(drug_names)>0) {
    		ith_cluster_drugs <- subset(ith_cluster_drugs, Drug.name %in% drug_names)
    		fdrs <- ith_cluster_drugs$FDR
    		p_values <- ith_cluster_drugs$P.value
    		
  	temp <- data.frame(
  		drug = drug_names, 
  		cluster = i,
  		cluster_prop = cluster_prop[i],
  		p_value = p_values,
  		fdr = fdrs,
  		row.names = NULL
  	)
    		drug_list <- rbind(drug_list, temp)
    	}
  }
  drug_list <- unique(drug_list)
  drug_list$weighted_prop <- drug_list$cluster_prop*(-log10(drug_list$fdr))
  drug_list[is.na(drug_list)] <- 0
  
  drug_coverage <- tapply(drug_list$weighted_prop, drug_list$drug, sum)
  drugs <- rownames(drug_coverage)
  
  # Combine cluster spesific p-values of drugs
  if(length(unique(names(cluster_drugs)))>1){
     	combined_p_values <- tapply(drug_list$p_value, drug_list$drug, CombineP)
  }else{
    	combined_p_values <- drug_list$p_value
    	names(combined_p_values) <- drug_list$drug
  }
  
  # Cell line information
  cell_lines <- subset(cell_data, primary_site == tissue)$cell_id
  
  # Load drugs metadata for GSE92742 and subset it to tissue of interest and 
  # drugs of interest
  drug_metadata_92742 <- col_meta_GSE92742[, c("sig_id", "pert_iname")]
  row.names(drug_metadata_92742) <- drug_metadata_92742$sig_id
  idx <- which(col_meta_GSE92742$cell_id %in% cell_lines & 
  		 col_meta_GSE92742$pert_iname %in% drugs)
  sig_ids <- col_meta_GSE92742$sig_id[idx]
  drug_metadata_92742 <- drug_metadata_92742[sig_ids, ]
  
  # Load drug response for GSE92742
  exprs <- as.data.frame(parse_gctx(gse92742_gctx_path, cid=sig_ids)@mat)
  treatments <- colnames(exprs)
  exprs$gene_id <- row.names(exprs)
  tmp <- merge(exprs, gene_meta, by.x="gene_id", by.y="pr_gene_id")
  drug_responses_92742 <- tmp[, c("pr_gene_symbol", treatments)]
  
  
  
  # Load drugs metadata for GSE70138 and subset it to tissue of interest and 
  # drugs of interest
  # drug_metadata_70138 <- col_meta_GSE70138[, c("sig_id", "pert_iname")]
  # row.names(drug_metadata_70138) <- drug_metadata_70138$sig_id
  # # idx <- which(col_meta_GSE70138$cell_id %in% cell_lines & 
  # # 		 col_meta_GSE70138$pert_iname %in% drugs)
  # idx <- which(col_meta_GSE70138$cell_id %in% cell_lines & 
  # 		 col_meta_GSE70138$pert_iname %in% drugs)
  # sig_ids <- col_meta_GSE70138$sig_id[idx]
  # drug_metadata_70138 <- drug_metadata_70138[sig_ids, ]
  # 
  # # Load drug response for GSE70138
  # exprs <- as.data.frame(parse_gctx(gse70138_gctx_path, cid=sig_ids)@mat)
  # treatments <- colnames(exprs)
  # exprs$gene_id <- row.names(exprs)
  # tmp <- merge(exprs, gene_meta, by.x="gene_id", by.y="pr_gene_id")
  # drug_responses_70138 <- tmp[, c("pr_gene_symbol", treatments)]
  
  
  
  # drug_responses <- merge(drug_responses_92742, drug_responses_70138, 
  # 					by="pr_gene_symbol")
  # row.names(drug_responses) <- drug_responses[, 1]
  # drug_responses <- drug_responses[, -1]
  # drug_metadata <- rbind(drug_metadata_92742, drug_metadata_70138)
  
  
  drug_responses <- drug_responses_92742
  row.names(drug_responses) <- drug_responses[, 1]
  drug_responses <- drug_responses[, -1]
  drug_metadata <- drug_metadata_92742
  
  
  
  
  # Find DEGs that are common to all clusters
  common_degs <- list()
  for (i in names(cluster_degs)) {
  	ith_cluster_degs <- cluster_degs[[i]]
    	ith_cluster_degs <- subset(ith_cluster_degs, adj.P.Val < 0.05)
  if (length(ith_cluster_degs) > 0) {
    	common_degs[[i]] <- rownames(ith_cluster_degs)
  }
  }
  common_degs <- Reduce(intersect, common_degs)
  
  # Combine cluster specific DEG scores into a matrix
  deg_scores <- data.frame()
  for (i in names(cluster_degs)) {
  	ith_cluster_degs <- cluster_degs[[i]]
  	if (nrow(deg_scores) == 0) {
  		deg_scores <- data.frame(score = ith_cluster_degs[common_degs, "score"])
  	} else {
  	    tmp <- data.frame(score = ith_cluster_degs[common_degs,"score"])
      	deg_scores <- cbind(deg_scores, tmp)
     }
  }
  deg_scores <- as.matrix(deg_scores)
  row.names(deg_scores) <- common_degs
  
  deg_scores_mean <- apply(deg_scores, 1, mean)
  names(deg_scores_mean) <- common_degs
  
  # Calculate drug score
  drug_scores <- list()
  for (drug in drugs) {
  # Get response from CMap
  treatments <- subset(drug_metadata, pert_iname == drug)$sig_id
  if (length(treatments) > 1) {
  	curr_drug_response <- drug_responses[, treatments]
  	mean_response <- apply(curr_drug_response, 1, mean)
  } else {
  	curr_drug_response <- drug_responses[, treatments]
  	mean_response <- curr_drug_response
  }
  
  drug_stats <- drug_list[drug_list$drug == drug, ]
  drug_score <- 0
  for (i in names(cluster_degs)) {
  	cluster_prop <- drug_stats[drug_stats$cluster == i, "cluster_prop"]
  	fdr <- drug_stats[drug_stats$cluster == i, "fdr"]
  	p_value <- drug_stats[drug_stats$cluster == i, "p_value"]
  
  	ith_cluster_degs <- cluster_degs[[i]]
    		ith_cluster_degs <- subset(ith_cluster_degs, adj.P.Val < 0.05)
  
  	treatable_degs <- intersect(row.names(ith_cluster_degs), names(mean_response))
  	if (length(treatable_degs > 0)) {
  		deg_scores <- ith_cluster_degs[treatable_degs, "score"]
  
  		treated_degs <- -deg_scores*mean_response[treatable_degs]
  		treated_degs <- treated_degs[which(treated_degs > 0)]
  
  		treated_degs_ratio <- length(treated_degs)/length(treatable_degs)
  		drug_score <- drug_score +
  			(cluster_prop/100)*(-log10(fdr))*treated_degs_ratio
  	}
    }
  
  drug_scores[[drug]] <- drug_score
  }
  drug_scores <- t(as.data.frame(drug_scores))
  
    out <- data.frame(
  	Drug.therapeutic.score = drug_scores,
  	P.value = combined_p_values[drugs],
  	FDR = p.adjust(combined_p_values[drugs], method = "BH")
  )
  return(out)

}


```


```{r}
# Change the following two lines with the paths on your computer
gse92742_gctx_path <- "/data/jp2079/KST_Data/AsgardData/GSE92742_Broad_LINCS_Level5_COMPZ.MODZ_n473647x12328.gctx"
gse70138_gctx_path <- "/data/jp2079/KST_Data/AsgardData/GSE70138_Broad_LINCS_Level5_COMPZ_n118050x12328_2017-03-06.gctx"

cell_metadata <- SC.integrated@meta.data
cell_metadata$cluster <- SC.integrated@meta.data$celltype

Case=c("PDX-110","PDX-332")

Drug.score <- DrugScore_jaesub(cell_metadata[,c('sample','cluster')], cluster_degs = Gene.list, 
                        cluster_drugs = Drug.ident.res, tissue = tissue, 
                        case = Case, gse92742_gctx_path = gse92742_gctx_path, 
                        gse70138_gctx_path = gse70138_gctx_path)

```


```{r}
cell_metadata_kst <- data.frame(
  sample = c("sampleA", "sampleA","sampleA","sampleA","sampleA","sampleA","sampleA",
               "sampleB","sampleB","sampleB","sampleB","sampleB","sampleB","sampleB"),   # sample 열에 "sampleA"를 3번 반복
  cluster = c("C1", "C2", "C3", 'C4', 'C5', 'C6', 'C7',
              "C1", "C2", "C3", 'C4', 'C5', 'C6', 'C7') # cluster 열에 "C1", "C2", "C3" 값 지정
)

Case=c("sampleA", 'sampleB')
# Case=c("PDX-110","PDX-332")

Drug.score.kst <- DrugScore_jaesub(cell_metadata_kst, cluster_degs = Gene.list.kst, 
                        cluster_drugs = Drug.ident.res, tissue = tissue, 
                        case = Case, gse92742_gctx_path = gse92742_gctx_path, 
                        gse70138_gctx_path = gse70138_gctx_path)

write.csv(Drug.score.kst, '/home/jp2079/KST_project/Drug_score_KST.csv')
```

```{r}
#Select drug using drug socre
# library(Hmisc)
# Final.drugs<-subset(Drug.score,Drug.therapeutic.score>quantile(Drug.score$Drug.therapeutic.score, 0.99,na.rm=T) & FDR <0.05)
# 

#Select drug for individual clusters
# Final.drugs<-TopDrug(SC.integrated=SC.integrated,
#                    Drug.data=Drug.ident.res,
#                    Drug.FDR=0.1,
#                    FDA.drug.only=TRUE,
#                    Case=Case.samples,
#                    DrugScore=FALSE
# )
```

Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Ctrl+Alt+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Ctrl+Shift+K* to preview the HTML file).

The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.
